devices/usbc/ctrl.c

```diff
@@ -0,0 +1,400 @@
+
+	base_addr = ((u32)ctrl_common.dc->endptqh[qh].head & ~((ctrl_common.dc->endptqh[qh].size * sizeof(dtd_t)) - 1));
+
+	ret = ctrl_common.dc->endptqh[qh].tail++;
+	ctrl_common.dc->endptqh[qh].tail = (dtd_t *)(base_addr | ((u32)ctrl_common.dc->endptqh[qh].tail & (((ctrl_common.dc->endptqh[qh].size) * sizeof(dtd_t)) - 1)));
+
+	return ret;
+}
+
+
+static int ctrl_buildDtd(dtd_t *dtd, u32 paddr, u32 size)
+{
+	int i = 0;
+	int tempSize = size;
+
+	if (size > USB_BUFFER_SIZE)
+		return -1;
+
+	dtd->dtd_next = 1;
+	dtd->dtd_token = size << 16;
+	dtd->dtd_token |= 1 << 7;
+
+	/* TODO: allow to use additional dtd within the same dQH */
+	while (tempSize > 0) {
+		dtd->buff_ptr[i++] = paddr;
+		tempSize -= 0x1000;
+		paddr += 0x1000;
+	}
+
+	return ERR_NONE;
+}
+
+
+dtd_t *ctrl_execTransfer(int endpt, u32 paddr, u32 sz, int dir)
+{
+	int shift;
+	u32 offs;
+	volatile dtd_t *dtd;
+	
+	int qh = (endpt << 1) + dir;
+
+	dtd = ctrl_getDtd(endpt, dir);
+
+	ctrl_buildDtd((dtd_t *)dtd, paddr, sz);
+
+	shift = endpt + ((qh & 1) ? 16 : 0);
+	offs = (u32)dtd & (((ctrl_common.dc->endptqh[qh].size) * sizeof(dtd_t)) - 1);
+
+	ctrl_common.dc->endptqh[qh].dtd_next = (ctrl_common.dc->endptqh[qh].base + offs) & ~1;
+	ctrl_common.dc->endptqh[qh].dtd_token &= ~(1 << 6);
+	ctrl_common.dc->endptqh[qh].dtd_token &= ~(1 << 7);
+
+	while ((*(ctrl_common.dc->base + endptprime) & (1 << shift)))
+		;
+
+	*(ctrl_common.dc->base + endptprime) |= 1 << shift;
+
+	/* prime the endpoint and wait for it to prime */
+	while ((*(ctrl_common.dc->base + endptprime) & (1 << shift)))
+		;
+	*(ctrl_common.dc->base + endptprime) |= 1 << shift;
+	while (!(*(ctrl_common.dc->base + endptprime) & (1 << shift)) && (*(ctrl_common.dc->base + endptstat) & (1 << shift)))
+		;
+
+	/* wait to finish transaction */
+	while (DTD_ACTIVE(dtd) && !DTD_ERROR(dtd))
+		;
+
+	return (dtd_t *)dtd;
+}
+
+
+int ctrl_hfIrq(void)
+{
+	int endpt = 0;
+
+	if ((ctrl_common.dc->setupstat = *(ctrl_common.dc->base + endptsetupstat)) & 0x1) {
+		/* trip winre set */
+		while (!((ctrl_common.dc->setupstat >> endpt) & 1))
+			endpt++;
+		do {
+			*(ctrl_common.dc->base + usbcmd) |= 1 << 13;
+			low_memcpy(&ctrl_common.dc->setup, (void *)ctrl_common.dc->endptqh[endpt].setup_buff, sizeof(usb_setup_packet_t));
+		} while (!(*(ctrl_common.dc->base + usbcmd) & 1 << 13));
+
+		*(ctrl_common.dc->base + endptsetupstat) |= 1 << endpt;
+		*(ctrl_common.dc->base + usbcmd) &= ~(1 << 13);
+		*(ctrl_common.dc->base + usbsts) |= 1;
+
+		ctrl_common.dc->endptqh[0].head = ctrl_common.dc->endptqh[0].tail;
+		ctrl_common.dc->endptqh[1].head = ctrl_common.dc->endptqh[1].tail;
+
+		while (*(ctrl_common.dc->base + endptsetupstat) & 1);
+		desc_setup(&ctrl_common.dc->setup);
+	}
+
+	return 1;
+}
+
+
+int ctrl_lfIrq(void)
+{
+	if ((*(ctrl_common.dc->base + usbsts) & 1 << 6)) {
+
+		*(ctrl_common.dc->base + endptsetupstat) = *(ctrl_common.dc->base + endptsetupstat);
+		*(ctrl_common.dc->base + endptcomplete) = *(ctrl_common.dc->base + endptcomplete);
```

Comment:
It is correct?

Reply:
This action is performed when USB reset state is received. 
Registers `endptsetupstat` and `endptcomplete` contain information about active transfers on specific endpoints. After reset, all active transfers should be cleared. 
According to documentation: _Writing one clears the corresponding bit in this register_. For this reason the same value is written to register to clear its state.